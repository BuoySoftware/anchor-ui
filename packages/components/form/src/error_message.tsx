import { WarningCircle } from "iconoir-react";
import { Box, BoxProps, Flex } from "@buoysoftware/anchor-layout";
import { Body, Markdown } from "@buoysoftware/anchor-typography";
import { useTheme } from "@buoysoftware/anchor-theme";

interface OwnProps {
  errors: string[];
  parentId: string;
}

type ErrorMessageProps = OwnProps & BoxProps;

export const ErrorMessage: React.FC<ErrorMessageProps> = ({
  errors,
  parentId,
  ...props
}): React.ReactElement => {
  const { colors } = useTheme();
  const id = `error-${parentId}`;

  return (
    <Flex
      bg="background.criticalSubdued"
      borderRadius="4px"
      data-testid={id}
      id={id}
      py="xxs"
      px="xs"
      role="alert"
      {...props}
    >
      <WarningCircle color={colors.text.critical} width={16} height={16} />
      <Box ml="xs">
        {errors.map((error, index) => (
          <Body as="div" size="m" key={`error-${index}`} color="text.critical">
            <Markdown>{error}</Markdown>
          </Body>
        ))}
      </Box>
    </Flex>
  );
};
