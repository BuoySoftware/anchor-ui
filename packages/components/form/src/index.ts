export * from "./checkbox";
export * from "./checkbox_field";
export * from "./error_message";
export * from "./form";
export * from "./form_base_errors";
export * from "./form_button";
export * from "./form_scope_provider";
export * from "./input";
export * from "./label";
export * from "./radio_field";
export * from "./select";
export * from "./select_field";
export * from "./text_area";
export * from "./text_field";
export * from "./use_conditional_field";
export * from "./use_form_field";

export { default as anchorFormsEn } from "./i18n/anchorForms.en.json";
