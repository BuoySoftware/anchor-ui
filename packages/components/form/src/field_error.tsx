import compact from "lodash/compact";
import flatten from "lodash/flatten";
import { FieldError as HookFormFieldError } from "react-hook-form";
import snakeCase from "lodash/snakeCase";
import { BoxProps } from "@buoysoftware/anchor-layout";
import { TFunction } from "i18next";
import { useTranslation } from "react-i18next";

import { useFormScope } from "./form_scope_provider";
import { ErrorMessage } from "./error_message";

export type InputType =
  | "checkbox"
  | "collectionCheckbox"
  | "hidden"
  | "phoneNumber"
  | "radio"
  | "select"
  | "text";

interface OwnProps {
  error?: HookFormFieldError;
  name?: string;
  inputId: string;
  inputType?: InputType;
}

export type FieldErrorProps = OwnProps & BoxProps;

/**
 * Look up an error message in the i18next translations. This utilizes the
 * i18next fallback mechanism for determining who to translate the error message.
 *
 * The order is as follows:
 *
 * 1. An error message in a custom provided namespace via tNamespace scoped to
 *    forms.errors and then the form scope, field name, and error typea.
 *
 * 2. An error message in the default forms namespace scoped to the form scope,
 *    field name and error type.
 *
 * 3. An error message in default forms namespace for the given input type and
 *    error type.
 *
 * 4. An error message in the anchorForms namespace for the given input type and
 *    error type.
 *
 * 5. An error message in the forms namespace for the given error type
 *
 * 6. An error message in the anchorForms namespace for the given error type
 */
function lookupErrorMessage(
  error: HookFormFieldError,
  scope: string,
  name: string,
  inputType: InputType,
  t: TFunction,
  tNamespace: string[]
): string {
  if (error.message !== undefined && error.message !== "") {
    return error.message;
  }

  const i18nName = snakeCase(name);
  const i18nErrorScope = snakeCase(scope);

  // These are translations scoped to the form's scope and field name.
  // The use of forms.errors is to look into the custom tNamespace if one was
  // provided. The second will look up in the forms default namespace.
  const fieldScopes = [
    `forms.errors.${i18nErrorScope}.${i18nName}.${error.type}`,
    `errors.${i18nErrorScope}.${i18nName}.${error.type}`,
  ];

  // Global scopes are used to look up defaults in the forms namespace or
  // the anchorForms namespace
  const globalScopes = [
    `errors.input_types.${snakeCase(inputType)}.${error.type}`,
    `errors.${error.type}`,
  ];

  const combinedGlobalScopes = tNamespace.flatMap((namespace: string) =>
    globalScopes.map((scope) => `${namespace}:${scope}`)
  );

  return t([...fieldScopes, ...combinedGlobalScopes]);
}

export const FieldError: React.FC<FieldErrorProps> = ({
  error,
  inputId,
  name,
  inputType = "text",
  ...elementProps
}): React.ReactElement | null => {
  const { scope, tNamespace = [] } = useFormScope();
  const namespace = compact(flatten([tNamespace, "forms", "anchorForms"]));
  const { t } = useTranslation(namespace);

  if (!error || !name) return null;

  const message = lookupErrorMessage(
    error,
    scope,
    name,
    inputType,
    t,
    namespace
  );

  return (
    <ErrorMessage
      errors={message.split("\n")}
      parentId={inputId}
      {...elementProps}
    />
  );
};
