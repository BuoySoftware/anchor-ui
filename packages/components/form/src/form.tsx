import kebabCase from "lodash/kebabCase";
import { useEffect, useState } from "react";
import {
  FieldValues,
  FormProvider,
  get,
  SubmitHandler,
  UseFormReturn,
} from "react-hook-form";
import { ErrorMessage } from "./error_message";
import { Box, BoxProps } from "@buoysoftware/anchor-layout";

import { FormScopeProvider, TNamespace } from "./form_scope_provider";

export type FormValidationErrors<T> = {
  [key in keyof T]: string[];
};

export type ErrorKeys<TFormValues> = Record<string, keyof TFormValues>;

interface OwnProps<TFormValues extends FieldValues> {
  "data-testid"?: string;
  id: string;
  scope: string;
  onSubmit?: SubmitHandler<TFormValues>;
  children: React.ReactNode;
  form: UseFormReturn<TFormValues>;
  errorKeys?: ErrorKeys<TFormValues>;
  mutationErrors?: Partial<FormValidationErrors<TFormValues>>;
  tNamespace?: TNamespace;
}

export type FormProps<TFormValues extends FieldValues> = OwnProps<TFormValues> &
  BoxProps;

export const Form = <TFormValues extends FieldValues>({
  "data-testid": testId,
  id,
  scope,
  children,
  errorKeys = {},
  form: { handleSubmit, ...form },
  onSubmit = () => undefined,
  mutationErrors = {},
  tNamespace,
  ...elementProps
}: FormProps<TFormValues>): React.ReactElement => {
  const [hasErrors, setHasErrors] = useState(false);

  useEffect(() => {
    if (hasErrors) {
      Object.entries(mutationErrors).forEach(([key, errors]) => {
        if (key !== "base") {
          const messages = errors?.join("\n");
          const inputKey = get(errorKeys, key, key);

          form.setError(inputKey as any, {
            type: "server",
            message: messages,
          });
        }
      });
    }
  }, [hasErrors, mutationErrors]);

  useEffect(() => {
    setHasErrors(!!Object.keys(mutationErrors).length);
  }, [mutationErrors]);

  const baseErrors = mutationErrors["base"] || [];

  return (
    <FormScopeProvider value={{ formId: id, baseErrors, scope, tNamespace }}>
      <FormProvider {...form} handleSubmit={handleSubmit}>
        <Box
          as="form"
          autoComplete="off"
          id={id}
          data-testid={testId || kebabCase(scope)}
          onSubmit={handleSubmit(onSubmit)}
          {...elementProps}
        >
          {children}
        </Box>
      </FormProvider>
    </FormScopeProvider>
  );
};
