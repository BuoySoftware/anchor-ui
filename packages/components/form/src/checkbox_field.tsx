import { Box, BoxProps } from "@buoysoftware/anchor-layout";
import { RegisterOptions, useFormContext } from "react-hook-form";

import { useFormField } from "./use_form_field";
import { InlineLabel } from "./inline_label";
import { Checkbox } from "./checkbox";
import { FieldError } from "./field_error";

interface OwnProps {
  name: string;
  value?: string;
  rules?: RegisterOptions;
}

type CheckboxFieldProps = OwnProps & BoxProps;

export type CheckboxFieldValue = "1" | false;

export const DEFAULT_CHECKED_VALUE: CheckboxFieldValue = "1";

export const CheckboxField: React.FC<CheckboxFieldProps> = ({
  name,
  value = DEFAULT_CHECKED_VALUE,
  rules,
  ...boxProps
}): React.ReactElement => {
  const { label, error, inputId } = useFormField({ name });
  const { register } = useFormContext();

  return (
    <Box
      as="fieldset"
      data-testid={`checkbox-field-${inputId}`}
      display="flex"
      flexDirection="column"
      mb="form.fieldGap"
      {...boxProps}
    >
      <InlineLabel
        data-testid={`label-${inputId}`}
        display="flex"
        alignItems="center"
        htmlFor={inputId}
      >
        <Checkbox
          aria-describedby={error ? `error-${name}` : undefined}
          aria-invalid={error ? "true" : "false"}
          data-testid={inputId}
          id={inputId}
          mr="xxs"
          value={value}
          {...register(name, { shouldUnregister: true, ...rules })}
        />
        {label}
      </InlineLabel>
      <FieldError
        error={error}
        inputId={inputId}
        inputType="checkbox"
        mt="s"
        name={name}
      />
    </Box>
  );
};
