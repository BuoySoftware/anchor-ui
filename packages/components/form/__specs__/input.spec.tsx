import { render, waitFor } from "../test_utils";
import { Search } from "iconoir-react";
import { theme } from "@buoysoftware/anchor-theme";

import { Input } from "../src";

describe("<Input />", () => {
  it("renders properly", () => {
    const { asFragment } = render(<Input id="example" name="example" />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders a disabled input properly", () => {
    const { asFragment } = render(
      <Input id="example" disabled={true} name="example" />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  context("unit", () => {
    it("renders a unit properly", () => {
      const { asFragment } = render(
        <Input id="example" unit="mL" name="example" />
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it("does not apply any spacing", () => {
      const { getByTestId } = render(
        <Input id="example" unit="mL" name="example" />
      );

      expect(getByTestId("input-unit")).toHaveStyleRule("margin-right", "0rem");
    });
  });

  context("prefix unit", () => {
    it("renders a prefix unit properly", () => {
      const { asFragment } = render(
        <Input id="example" unit="$" prefixUnit={true} name="example" />
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it("spaces the prefix unit properly", () => {
      const { getByTestId } = render(
        <Input id="example" unit="$" prefixUnit={true} name="example" />
      );

      expect(getByTestId("input-unit")).toHaveStyleRule(
        "margin-right",
        theme.space.xs
      );
    });
  });

  context("input with error", () => {
    it("applies a critical border", () => {
      const { getByTestId } = render(
        <Input id="example" error={true} name="example" />
      );

      expect(getByTestId("input-wrapper-example")).toHaveStyleRule(
        "border-color",
        theme.colors.border.critical
      );
    });
  });

  it("renders a readOnly input properly", () => {
    const { asFragment } = render(
      <Input id="example" readOnly={true} name="example" />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders with an icon properly", () => {
    const { asFragment } = render(
      <Input id="example" icon={<Search />} name="example" />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders a focus state properly", async () => {
    const { getByTestId, asFragment } = render(
      <Input id="example" name="example" />
    );

    await waitFor(() => getByTestId("input-example").focus());

    expect(asFragment()).toMatchSnapshot();
  });

  it("supports a custom onFocus callback", async () => {
    const onFocus = jest.fn();
    const { getByTestId } = render(
      <Input id="example" name="example" onFocus={onFocus} />
    );

    await waitFor(() => getByTestId("input-example").focus());

    expect(onFocus).toHaveBeenCalled();
  });
});
