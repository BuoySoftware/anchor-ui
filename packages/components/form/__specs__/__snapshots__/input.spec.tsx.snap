// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> prefix unit renders a prefix unit properly 1`] = `
<DocumentFragment>
  .c0 {
  box-sizing: border-box;
}

.c2 {
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4 {
  box-sizing: border-box;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c6 {
  box-sizing: border-box;
  color: #9CA1A5;
  margin-right: 0.5rem;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
}

.c5 {
  outline: none;
  background: none;
  border: none;
  color: inherit;
  fontsize: inherit;
  line-height: inherit;
  margin: 0;
  padding: 0;
  width: 100%;
  text-align: inherit;
  box-sizing: border-box;
}

.c1 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Segoe UI, Roboto,Helvetica Neue,sans-serif;
  display: inline-block;
  background-color: #FFFFFF;
  border: 1px solid;
  border-radius: 4px;
  color: #101112;
  font-size: 0.875rem;
  height: 40px;
  line-height: 1rem;
  margin: 0;
  padding: 0.75rem 0.75rem;
  width: 300px;
}

.c1 transition:"background 0.75s ease",
.c1 input::-webkit-input-placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c1 transition:"background 0.75s ease",
.c1 input::-moz-placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c1 transition:"background 0.75s ease",
.c1 input:-ms-input-placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c1 transition:"background 0.75s ease",
.c1 input::placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c1:hover {
  border-color: #46494D;
}

<div
    class="c0 c1"
    data-testid="input-wrapper-example"
  >
    <div
      class="c2 c3"
      display="flex"
    >
      <div
        class="c4"
      >
        <input
          class="c5"
          data-testid="input-example"
          id="example"
          name="example"
          value=""
        />
      </div>
      <div
        class="c6"
        color="text.tertiary"
        data-testid="input-unit"
      >
        $
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Input /> renders a disabled input properly 1`] = `
<DocumentFragment>
  .c0 {
  box-sizing: border-box;
}

.c2 {
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4 {
  box-sizing: border-box;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c5 {
  outline: none;
  background: none;
  border: none;
  color: inherit;
  fontsize: inherit;
  line-height: inherit;
  margin: 0;
  padding: 0;
  width: 100%;
  text-align: inherit;
  box-sizing: border-box;
}

.c1 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Segoe UI, Roboto,Helvetica Neue,sans-serif;
  display: inline-block;
  background-color: #FFFFFF;
  border: 1px solid;
  border-radius: 4px;
  color: #101112;
  font-size: 0.875rem;
  height: 40px;
  line-height: 1rem;
  margin: 0;
  padding: 0.75rem 0.75rem;
  width: 300px;
  background: #FAFBFC;
  border-color: #BABEC3;
  color: #9CA1A5;
}

.c1 transition:"background 0.75s ease",
.c1 input::-webkit-input-placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c1 transition:"background 0.75s ease",
.c1 input::-moz-placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c1 transition:"background 0.75s ease",
.c1 input:-ms-input-placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c1 transition:"background 0.75s ease",
.c1 input::placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c1:hover {
  border-color: #46494D;
}

.c1 input {
  -webkit-text-fill-color: #9CA1A5;
}

<div
    class="c0 c1"
    data-testid="input-wrapper-example"
    disabled=""
  >
    <div
      class="c2 c3"
      display="flex"
    >
      <div
        class="c4"
      >
        <input
          class="c5"
          data-testid="input-example"
          disabled=""
          id="example"
          name="example"
          value=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Input /> renders a focus state properly 1`] = `
<DocumentFragment>
  .c0 {
  box-sizing: border-box;
}

.c2 {
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4 {
  box-sizing: border-box;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c5 {
  outline: none;
  background: none;
  border: none;
  color: inherit;
  fontsize: inherit;
  line-height: inherit;
  margin: 0;
  padding: 0;
  width: 100%;
  text-align: inherit;
  box-sizing: border-box;
}

.c1 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Segoe UI, Roboto,Helvetica Neue,sans-serif;
  display: inline-block;
  background-color: #FFFFFF;
  border: 1px solid;
  border-radius: 4px;
  color: #101112;
  font-size: 0.875rem;
  height: 40px;
  line-height: 1rem;
  margin: 0;
  padding: 0.75rem 0.75rem;
  width: 300px;
  box-shadow: 0px 0px 0px 2px #FFFFFF, 0px 0px 0px 4px #5B91F4;
}

.c1 transition:"background 0.75s ease",
.c1 input::-webkit-input-placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c1 transition:"background 0.75s ease",
.c1 input::-moz-placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c1 transition:"background 0.75s ease",
.c1 input:-ms-input-placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c1 transition:"background 0.75s ease",
.c1 input::placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c1:hover {
  border-color: #46494D;
}

<div
    class="c0 c1"
    data-testid="input-wrapper-example"
  >
    <div
      class="c2 c3"
      display="flex"
    >
      <div
        class="c4"
      >
        <input
          class="c5"
          data-testid="input-example"
          id="example"
          name="example"
          value=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Input /> renders a readOnly input properly 1`] = `
<DocumentFragment>
  .c0 {
  box-sizing: border-box;
}

.c2 {
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4 {
  box-sizing: border-box;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c5 {
  outline: none;
  background: none;
  border: none;
  color: inherit;
  fontsize: inherit;
  line-height: inherit;
  margin: 0;
  padding: 0;
  width: 100%;
  text-align: inherit;
  box-sizing: border-box;
}

.c1 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Segoe UI, Roboto,Helvetica Neue,sans-serif;
  display: inline-block;
  background-color: #FFFFFF;
  border: 1px solid;
  border-radius: 4px;
  color: #101112;
  font-size: 0.875rem;
  height: 40px;
  line-height: 1rem;
  margin: 0;
  padding: 0.75rem 0.75rem;
  width: 300px;
  background: #FAFBFC;
  border-color: #BABEC3;
  color: #9CA1A5;
}

.c1 transition:"background 0.75s ease",
.c1 input::-webkit-input-placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c1 transition:"background 0.75s ease",
.c1 input::-moz-placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c1 transition:"background 0.75s ease",
.c1 input:-ms-input-placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c1 transition:"background 0.75s ease",
.c1 input::placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c1:hover {
  border-color: #46494D;
}

.c1 input {
  -webkit-text-fill-color: #9CA1A5;
}

<div
    class="c0 c1"
    data-testid="input-wrapper-example"
    readonly=""
  >
    <div
      class="c2 c3"
      display="flex"
    >
      <div
        class="c4"
      >
        <input
          class="c5"
          data-testid="input-example"
          id="example"
          name="example"
          readonly=""
          value=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Input /> renders properly 1`] = `
<DocumentFragment>
  .c0 {
  box-sizing: border-box;
}

.c2 {
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4 {
  box-sizing: border-box;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c5 {
  outline: none;
  background: none;
  border: none;
  color: inherit;
  fontsize: inherit;
  line-height: inherit;
  margin: 0;
  padding: 0;
  width: 100%;
  text-align: inherit;
  box-sizing: border-box;
}

.c1 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Segoe UI, Roboto,Helvetica Neue,sans-serif;
  display: inline-block;
  background-color: #FFFFFF;
  border: 1px solid;
  border-radius: 4px;
  color: #101112;
  font-size: 0.875rem;
  height: 40px;
  line-height: 1rem;
  margin: 0;
  padding: 0.75rem 0.75rem;
  width: 300px;
}

.c1 transition:"background 0.75s ease",
.c1 input::-webkit-input-placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c1 transition:"background 0.75s ease",
.c1 input::-moz-placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c1 transition:"background 0.75s ease",
.c1 input:-ms-input-placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c1 transition:"background 0.75s ease",
.c1 input::placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c1:hover {
  border-color: #46494D;
}

<div
    class="c0 c1"
    data-testid="input-wrapper-example"
  >
    <div
      class="c2 c3"
      display="flex"
    >
      <div
        class="c4"
      >
        <input
          class="c5"
          data-testid="input-example"
          id="example"
          name="example"
          value=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Input /> renders with an icon properly 1`] = `
<DocumentFragment>
  .c0 {
  box-sizing: border-box;
}

.c2 {
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4 {
  box-sizing: border-box;
  margin-right: 0.5rem;
}

.c5 {
  box-sizing: border-box;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c6 {
  outline: none;
  background: none;
  border: none;
  color: inherit;
  fontsize: inherit;
  line-height: inherit;
  margin: 0;
  padding: 0;
  width: 100%;
  text-align: inherit;
  box-sizing: border-box;
}

.c1 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Segoe UI, Roboto,Helvetica Neue,sans-serif;
  display: inline-block;
  background-color: #FFFFFF;
  border: 1px solid;
  border-radius: 4px;
  color: #101112;
  font-size: 0.875rem;
  height: 40px;
  line-height: 1rem;
  margin: 0;
  padding: 0.5rem 0.75rem;
  width: 300px;
}

.c1 transition:"background 0.75s ease",
.c1 input::-webkit-input-placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c1 transition:"background 0.75s ease",
.c1 input::-moz-placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c1 transition:"background 0.75s ease",
.c1 input:-ms-input-placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c1 transition:"background 0.75s ease",
.c1 input::placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c1:hover {
  border-color: #46494D;
}

<div
    class="c0 c1"
    data-testid="input-wrapper-example"
  >
    <div
      class="c2 c3"
      display="flex"
    >
      <div
        class="c4"
      >
        <svg
          color="currentColor"
          fill="none"
          height="1.5em"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          width="1.5em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M17 17l4 4M3 11a8 8 0 1016 0 8 8 0 00-16 0z"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </div>
      <div
        class="c5"
      >
        <input
          class="c6"
          data-testid="input-example"
          id="example"
          name="example"
          value=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Input /> unit renders a unit properly 1`] = `
<DocumentFragment>
  .c0 {
  box-sizing: border-box;
}

.c2 {
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4 {
  box-sizing: border-box;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c6 {
  box-sizing: border-box;
  color: #9CA1A5;
  margin-right: 0rem;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c5 {
  outline: none;
  background: none;
  border: none;
  color: inherit;
  fontsize: inherit;
  line-height: inherit;
  margin: 0;
  padding: 0;
  width: 100%;
  text-align: inherit;
  box-sizing: border-box;
}

.c1 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Segoe UI, Roboto,Helvetica Neue,sans-serif;
  display: inline-block;
  background-color: #FFFFFF;
  border: 1px solid;
  border-radius: 4px;
  color: #101112;
  font-size: 0.875rem;
  height: 40px;
  line-height: 1rem;
  margin: 0;
  padding: 0.75rem 0.75rem;
  width: 300px;
}

.c1 transition:"background 0.75s ease",
.c1 input::-webkit-input-placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c1 transition:"background 0.75s ease",
.c1 input::-moz-placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c1 transition:"background 0.75s ease",
.c1 input:-ms-input-placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c1 transition:"background 0.75s ease",
.c1 input::placeholder {
  opacity: 1;
  color: #9CA1A5;
  -webkit-text-fill-color: $({ theme}) => theme.colors.text.tertiary;
}

.c1:hover {
  border-color: #46494D;
}

<div
    class="c0 c1"
    data-testid="input-wrapper-example"
  >
    <div
      class="c2 c3"
      display="flex"
    >
      <div
        class="c4"
      >
        <input
          class="c5"
          data-testid="input-example"
          id="example"
          name="example"
          value=""
        />
      </div>
      <div
        class="c6"
        color="text.tertiary"
        data-testid="input-unit"
      >
        mL
      </div>
    </div>
  </div>
</DocumentFragment>
`;
